torch/_functorch/aot_autograd.py

```diff
@@ -1668,6 +1756,229 @@ def describe_input(i, aot_config):
     else:
         return f"input {i - aot_config.num_params_buffers}"
 
+# TODO: document this better
+@dataclass
+class CompiledRuntimeMetadata:
+    # TODO: This type / object should be cleaned up
+    synthetic_base_info: Optional[List[Union[int, Tuple[int, torch.Tensor]]]]
+    fw_metadata: ViewAndMutationMeta
+
+    def __post_init__(self):
+        self.num_outputs = len(self.fw_metadata.output_info)
+        self.num_outputs_non_aliased = len(
+            [x for x in self.fw_metadata.output_info if x.output_type == OutputType.non_alias]
+        )
+        self.num_outputs_aliased_to_inputs = len(
+            [
+                x
+                for x in self.fw_metadata.output_info
+                if x.output_type in [
+                    OutputType.alias_of_input,
+                    OutputType.is_input,
+                ]
+            ]
+        )
+        self.num_outputs_aliased_to_intermediates = len(
+            [
+                x
+                for x in self.fw_metadata.output_info
+                if x.output_type in [
+                    OutputType.alias_of_intermediate,
+                    OutputType.alias_of_intermediate_save_as_output,
+                    OutputType.alias_of_intermediate_base_is_user_output,
+                ]
+            ]
+        )
+        self.num_outputs_aliased = (
+            self.num_outputs_aliased_to_inputs + self.num_outputs_aliased_to_intermediates
+        )
+        self.num_mutated_data_inputs = len(
+            [x for x in self.fw_metadata.input_info if x.mutates_data]
+        )
+        self.num_mutated_metadata_inputs = len(
+            [
+                x
+                for x in self.fw_metadata.input_info
+                if x.mutates_metadata
+            ]
+        )
+        self.num_mutated_metadata_only_inputs = len(
+            [
+                x
+                for x in self.fw_metadata.input_info
+                if not x.mutates_data and x.mutates_metadata
+            ]
+        )
+        self.num_mutated_inputs = self.num_mutated_data_inputs + self.num_mutated_metadata_only_inputs
+#
+# The wrapper created by this function handles all of the runtime aliasing and mutation "epilogue" logic
+# that needs to run after the compiled function.
+#
+# This function accepts a trace_joint flag, indicating whether or not we're generating the runtime
+# epilogue for a forward-only inference graph, or for an autograd.Function.apply function.
+# This is because there are some minor differences in how we treat these cases at runtime:
+# - resize_() is currently handled in the inference case, but not fully handled in the autograd case.
+# - the autograd cases inserts TensorAlias wrapper objects for outputs that alias inputs
+def create_runtime_wrapper(
```

Comment:
What modifications were made here when it was moved?

Reply:
No functional changes. The only difference was that I gave it the extra argument, `keep_input_mutations` - and the only reason for that is because the debug asserts change slightly when you use `keep_input_mutations`.

The reason that there are no functional changes in the runtime wrapper, is because the wrapper goes off of `runtime_metadata.fw_metadata.mutated_inp_indices` to know which indices to apply a mutation with. And our trace-time logic has already ensured that data-only mutation indices don't show up in this list when `keep_input_mutations` is set. So the wrapper itself doesn't need to do anything different - it just blindly applies updates to the inputs based on that list of indices.